blake2: x86 coresimd

kazcw opened the pull request:
cf. #56

Replacement for simd + simd_opt features. Builds on stable rustc 1.27+

The coresimd feature overrides all other simd flags. Build with coresimd will fail for unsupported target platforms, e.g. non-x86.

Benched on a Xeon X5650:
sse2 only: faster than simd_opt
ssse3: slightly faster than simd_opt

Benched on a Haswell:
avx2: same speed as simd_opt

NB. the standard CI settings won't test this properly since it introduces a new feature

kazcw committed:
factor out compressor to reuse for blake2bp/sp

kazcw committed:
blake2bp/blake2sp

kazcw committed:
blake2b/s: simplify update

kazcw committed:
x86 coresimd

Replacement for simd + simd_opt. Builds on stable rustc 1.27+

The coresimd feature overrides all other simd flags. Build with coresimd
will fail for unsupported target platforms, e.g. non-x86.

Benched on a Xeon X5650:
sse2 only: faster than simd_opt
ssse3: slightly faster than simd_opt

Benched on a Haswell:
avx2: same speed as simd_opt

newpavlov commented:
Thank you! 

You can introduce a separate line for CI like was [done](https://github.com/RustCrypto/block-ciphers/blob/master/.travis.yml#L14) for testing AES-NI. I will try to review both of your blake2 PRs on this weekend.

kazcw committed:
test coresimd on travis

This will test the sse2 and probably ssse3 coresimd implementations. If
the buildbox happens to support avx2, it will also test the avx2
implementations.

tarcieri commented:
@kazcw are you interested in continuing to work on this PR, and if so, can you rebase?

kazcw commented:
@tarcieri As with #70, better to start from scratch at this point.
